### DISCUSSION TITLE
Using a Custom Transport with MetaMask

### DISCUSSION THREAD

**Discussion by mario-christopher** [wevm/viem#1223] [Question]
Hi I am trying to use Viem to read a contract.
I am using a custom `transport` with the Provider (window.ethereum) from Metamask, but unable to get my code to work.
Here is my code sample and the error.

```
      const publicClient = createPublicClient({
        chain: mainnet,
        transport: custom((window as any).ethereum),
      });

      const blockNumber = await publicClient.getBlockNumber()
      console.log('BLOCKNUMBER', blockNumber);

      const contract = getContract({
        address: '0xFBA3912Ca04dd458c843e2EE08967fC04f3579c2',
        abi: wagmiAbi,
        publicClient,
      });

     /// This call fails with the error: Error#1: JSON-RPC responses must be JSON serializable objects.
      const totalSupply = await publicClient.readContract({      
        ...contract,
        functionName: 'totalSupply',
      });
      console.log('TOTALSUPPLY', totalSupply);
```

Is this the correct usage with a custom Transport object ?
Thanks.

**Comment by mario-christopher** [comment]
I also tried the above with `createWalletClient( ... ).extend(publicActions)` and faced the same issue.

**Comment by jxom** [comment]
works for me: https://stackblitz.com/edit/viem-getting-started-qtccnn?file=index.ts

**Reply by jxom** [reply]
(please fork and repro your issue)

**Comment by mario-christopher** [comment]
Hi @jxom 
Yes, the code for `client.getBlockNumber()` works.
But if you see my code sample, the error occurs when interacting with a Contract using the `await publicClient.readContract()` method.

Can you please try a code sample to interact with a Contract using `window.ethereum`.

Thanks.

**Reply by jxom** [reply]
Can you fork my repro and reproduce it? Thanks
